crates/oxc_span/src/cmp.rs

```diff
@@ -0,0 +1,158 @@
+    #[inline]
+    #[must_use]
+    fn content_ne(&self, other: &Rhs) -> bool {
+        !self.content_eq(other)
+    }
+}
+
+impl ContentEq for () {
+    #[inline]
+    fn content_eq(&self, _other: &()) -> bool {
+        true
+    }
+
+    #[inline]
+    fn content_ne(&self, _other: &()) -> bool {
+        false
+    }
+}
+
+/// Blanket implementation for references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mutable references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for [Option] types
+impl<T: ContentEq> ContentEq for Option<T> {
+    #[inline]
+    fn content_eq(&self, other: &Self) -> bool {
+        // NOTE: based on the standard library
+        // Spelling out the cases explicitly optimizes better than
+        // `_ => false`
+        #[allow(clippy::match_same_arms)]
+        match (self, other) {
+            (Some(lhs), Some(rhs)) => lhs.content_eq(rhs),
+            (Some(_), None) => false,
+            (None, Some(_)) => false,
+            (None, None) => true,
+        }
+    }
+}
+
+/// Blanket implementation for [oxc_allocator::Box] types
+impl<'a, T: ContentEq> ContentEq for oxc_allocator::Box<'a, T> {
+    #[inline]
+    fn content_eq(&self, other: &Self) -> bool {
+        self.as_ref().content_eq(other.as_ref())
+    }
+}
+
+/// Blanket implementation for [oxc_allocator::Vec] types
+///
+/// # WARNING
+///
+/// This implementation is slow compared to `[PartialEq]`, Consider comparing the 2 vectors using
+/// that when one is implemented. This implementation takes triple the time the `PartialEq` would take.
```

Comment:
Why is `ContentEq` so much slower than `PartialEq`?

Reply:
This is the extreme case, Most of the time it doesn't perform that much worse however this is a thing.

I should update this comment to better explain it, Inherintly `ContentEq` does more work. So when a type implements `PartialEq` we should always prefer it as it would more or less optimize down to something similar to a `memcmp` + a length check.

Otherwise, for simple types, these 2 would result in almost identical assembly when optimization is on.

https://godbolt.org/z/KzaTxTd7M

Reply:
Interesting. Yes you're right - for native primitive types, `PartialEq` is faster.

But if the type being compared is even a native primitive wrapped in a struct, then `ContentEq` and `PartialEq` seem to produce the same assembly: https://godbolt.org/z/54on5sMWc

NB: Even adding `#[derive(Clone, Copy, Eq)]` on `WrappedU32` doesn't make it kick in the memcmp optimization.

I guess `std` has specialized impls for `PartialEq` of slices of primitives types, but they don't generalize to slices of `Copy` types in general. I don't think we could do that even if we wanted to, because specialization requires nightly features, if I understand right.

Luckily for us, the AST doesn't contain any `Vec<u32>`s (or similar), so for our purposes I think `ContentEq` and `PartialEq` are going to be equally fast.